### PR TITLE
GitHub dependency support

### PR DESCRIPTION
**Pull request by gilligan**
This PR provides support for dependencies from GitHub (as `dependencies` or `devDependencies`):

- GitHub dependencies are replaced with `tar` nix store paths in package.json and package-lock.json
- GitHub urls in `requires` in the `package-lock.json` also get replaced
- URLs in `requires` for whatever reason need to be *without* leading `file://` unlike everywhere else
- Added some tests cases 

All of this seems to work for a non-trivial project that i am using this with.

### CODE REVIEW

File: internal.nix [hunk 1]
```diff
@@ -29,50 +93,110 @@ rec {
...
       isBundled = spec ? bundled && spec.bundled == true;
+      hasGitHubRequires = spec: (spec ? requires) && (lib.any (x: lib.hasPrefix "github:" x) (lib.attrValues spec.requires));
     in
-    if lib.hasPrefix "github:" (spec.from or "") || lib.hasPrefix "github:" (spec.source or "") then
-      throw "[npmlock2nix] The given package-lock.json contains sources that refer to GitHub. The source spec is often not precise enough to be translated into a (reliable) Nix git fetch invocation."
-    else
-      (spec // lib.optionalAttrs (!isBundled) ({
-        resolved = "file://" + (toString (makeSource name spec));
-      }) // lib.optionalAttrs (spec ? dependencies) {
+    (spec //
```
Comment (andir) [inline internal.nix:135]: Ugh this looks like a lot to take in. Can we reformat this more readable?

File: internal.nix [hunk 2]
```diff
@@ -139,17 +265,15 @@ rec {
               )}
 
             if grep -I -q -r '/bin/' .; then
-              source $TMP/preinstall-env
+              source $stdenv/setup
```
Comment (andir) [inline internal.nix:268]: Why did you revert this change? This was consciously done to speed up the repetitive sourcing. IIRC it provided a >2x build speed improvement (multiple minutes to sub one minute) on big projects.

File: internal.nix [hunk 3]
```diff
@@ -172,14 +296,15 @@ rec {
 
         postPatch = ''
           ln -sf ${patchedLockfile packageLockJson} package-lock.json
+          ln -sf ${patchedPackagefile packageJson} package.json
         '';
 
         buildPhase = ''
           runHook preBuild
           mkdir -p node_modules/.hooks
-          declare -pf > $TMP/preinstall-env
```
Comment (andir) [inline internal.nix:180]: This should also be restored.

File: internal.nix [hunk 4]
```diff
@@ -172,14 +296,15 @@ rec {
...
           ln -s ${preinstall_node_modules}/node_modules/.hooks/prepare node_modules/.hooks/preinstall
-          npm install --offline --nodedir=${nodeSource nodejs}
+          export HOME=.
+          npm i --offline --nodedir=${nodeSource nodejs}
```
Comment (andir) [inline internal.nix:307]: I think `npm install` is better in this case. The shortcuts should be used on the CLI only. For code (that is mostly read) it should be the "verbose" version.

File: tests/parse-github-ref.nix [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+      expected = {
+        parts = [ "github" [ ] "foo" [ ] "bar" [ ] "939360f9d1bafa9019b6ff8739495c6c9101c4a1" ];
+        org = "foo";
+        repo = "bar";
+        rev = "939360f9d1bafa9019b6ff8739495c6c9101c4a1";
+      };
+    };
+
+  # It would be nice if there was a way to test the failure
+  # case but Unfortunately we can't catch exceptions...
```
Comment (andir) [inline tests/parse-github-ref.nix:18]:
```suggestion
  # case but unfortunately we can't catch exceptions...
```

File: tests/patch-lockfile.nix [hunk 6]
```diff
@@ -1,6 +1,23 @@
...
+          version = "github:tmcw/leftpad#db1442a0556c2b133627ffebf455a78a1ced64b9";
+          from = "github:tmcw/leftpad#db1442a0556c2b133627ffebf455a78a1ced64b9";
+          integrity = "sha512-8/UvHFG90J4O4QNRzb0jB5Ni1QuvuB7XFTLfDMQnCzAsFemF29VKnNGUESFFcSP/r5WWh/PMe0YRz90+3IqsUA==";
+          requires = {
+            libxmljs = "github:znerol/libxmljs#0517e063347ea2532c9fdf38dc47878c628bf0ae";
+          };
+        }
+        ).requires.libxmljs;
+      in
+      lib.hasPrefix "/nix/store/" libxmljsUrl;
```
Comment (andir) [inline tests/patch-lockfile.nix:17]:
```suggestion
      lib.hasPrefix builtins.storeDir libxmljsUrl;
```

File: tests/patch-lockfile.nix [hunk 7]
```diff
@@ -71,6 +88,15 @@ testLib.runTests {
     expected = 1;
   };
 
+  testPatchLockfileTurnsGitHubUrlsIntoStorePaths = {
+    expr =
+      let
+        leftpad = (npmlock2nix.internal.patchLockfile ./examples-projects/github-dependency/package-lock.json).dependencies.leftpad;
+      in
+      lib.hasPrefix "file:///nix/store/" leftpad.version;
```
Comment (andir) [inline tests/patch-lockfile.nix:96]:
```suggestion
      lib.hasPrefix ("file://" + builtins.storeDir) leftpad.version;
```

File: tests/patch-packagefile.nix [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+let
+  i = npmlock2nix.internal;
+in
+(testLib.runTests {
+  testTurnsGitHubRefsToStorePaths = {
+    expr =
+      let
+        leftpad = (npmlock2nix.internal.patchPackagefile ./examples-projects/github-dependency/package.json).dependencies.leftpad;
+      in
+      lib.hasPrefix "file:///nix/store" leftpad;
```
Comment (andir) [inline tests/patch-packagefile.nix:11]: Same as above.

File: tests/patch-packagefile.nix [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+    expected = true;
+  };
+  testHandlesDevDependencies = {
+    expr =
+      let
+        leftpad = (npmlock2nix.internal.patchPackagefile ./examples-projects/github-dev-dependency/package.json).devDependencies.leftpad;
+      in
+      lib.hasPrefix "file:///nix/store" leftpad;
```
Comment (andir) [inline tests/patch-packagefile.nix:19]: Same as above.

File: tests/shell.nix [hunk 10]
```diff
@@ -7,6 +7,8 @@ testLib.runTests {
       custom_nodejs = symlinkJoin {
         name = "custom-nodejs";
         paths = [ nodejs ];
+        version = "12.8.3";
```
Comment (andir) [inline tests/shell.nix:10]: What is this version string used for? Is it the nodejs version we are using?

**Review by andir — COMMENTED** [review]
Looks good besides those few comments that I made.

File: internal.nix (hunk 2 above)
Comment (gilligan) [reply internal.nix:268]: uh.. that was totally not intentional. I think i made a mistake during a rebase! I'll fix it

File: internal.nix (hunk 4 above)
Comment (gilligan) [reply internal.nix:307]: sure, fine by me

File: tests/shell.nix (hunk 10 above)
Comment (gilligan) [reply tests/shell.nix:10]: https://github.com/tweag/npmlock2nix/blob/github-experiment/internal.nix#L191

The test won't work without this because it is used by the `nodeSource` function.

File: internal.nix [hunk 11]
```diff
@@ -29,50 +93,105 @@ rec {
...
   patchDependency = name: spec:
-    assert (builtins.typeOf name != "string") -> builtins.throw "Name of dependency ${toString name} must be a string";
-    assert (builtins.typeOf spec != "set") -> builtins.throw "Spec of dependency ${toString name} must be a set";
+    assert (builtins.typeOf name != "string") ->
+      throw "[npmlock2nix] Name of dependency ${toString name} must be a string";
+    assert (builtins.typeOf spec != "set") ->
+      throw "[npmlock2nix] pec of dependency ${toString name} must be a set";
     let
       isBundled = spec ? bundled && spec.bundled == true;
+      hasGitHubRequires = spec: (spec ? requires) && (lib.any (x: lib.hasPrefix "github:" x) (lib.attrValues spec.requires));
```
Comment (andir) [inline internal.nix:133]: Much better! Thank you :+1:

**Review by andir — APPROVED** [review]
Looks good to me. I can still build all the projects I was able to build previously.
